Improve HTTPS detection accuracy in Site Health

felixarntz opened the pull request:
This PR addresses follow-up feedback to https://core.trac.wordpress.org/changeset/49904 from https://github.com/WordPress/wordpress-develop/pull/870#issuecomment-763015259 (that PR is related, but targets another ticket).

It also renames a function as suggested by [this comment](https://core.trac.wordpress.org/ticket/47577#comment:31).

Trac ticket: https://core.trac.wordpress.org/ticket/47577

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -1498,6 +1498,10 @@ public function get_test_is_in_debug_mode() {
 	public function get_test_https_status() {
+		// Enforce fresh HTTPS detection results. This is normally invoked by using cron, but for Site Health it should
+		// always rely on the latest results.
+		wp_update_https_detection_errors();
```
Great!

TimothyBJacobs replied:
Should this be an async test now so we don't slow down page load?

adamsilverstein commented on src/wp-includes/https-detection.php:
```diff
@@ -9,28 +9,53 @@
-	}
-
-	return true;
+	return 'https' === wp_parse_url( $site_url, PHP_URL_SCHEME );
```
nice

adamsilverstein commented:
👍🏼 Overall code changes look great; i'll give this a test locally again as well.

TimothyBJacobs commented:
How do you feel about making `wp_update_https_detection_errors` shortcircuitable? The reason I ask is I maintain a plugin that functions similarly, but uses an external server to ping the site to check for HTTPS to account for instances where a loopback request results in a false negative. Ideally, we'd be able to override the check to use the more accurate detection, but still keep the user in Site Health.

felixarntz commented:
@TimothyBJacobs Great suggestion - I'll add that tomorrow.

adamsilverstein commented:
@felixarntz I did some testing locally and took some screenshots. Noticed a couple of tiny issues:

Accessing wp-admin via http (with my invalid "self signed" SSL cert), I noticed an issue where the title is getting some extraneous encoding:

![image](https://user-images.githubusercontent.com/2676022/106097528-b337d700-60f4-11eb-86ea-95b4755c0224.png)

When I changed my WordPress Address to **http** I got a different message, looks good:

![image](https://user-images.githubusercontent.com/2676022/106097691-01e57100-60f5-11eb-9ae5-851b9f4b27fa.png)

Site address only set to **http**, looks good:
![image](https://user-images.githubusercontent.com/2676022/106097819-33f6d300-60f5-11eb-9dfe-b1fcb74c0a69.png)

Both set to **http**, switched to accessing wp-admin via http:

There is a slight problem here, the space is missing after the work "There":

![image](https://user-images.githubusercontent.com/2676022/106098031-918b1f80-60f5-11eb-94ff-43ab90829349.png)

felixarntz commented:
@adamsilverstein @TimothyBJacobs Addressed your feedback.

Here are some screenshots of the different states.

**HTTPS not set up, not supported**
![Screen Shot 2021-01-28 at 12 19 39 PM](https://user-images.githubusercontent.com/3531426/106194324-8373e800-6163-11eb-8686-15fb600a3455.png)

**HTTPS not set up, but supported**
![Screen Shot 2021-01-28 at 12 16 37 PM](https://user-images.githubusercontent.com/3531426/106194376-9686b800-6163-11eb-8189-d3af7c65c4b6.png)

**HTTPS not set up, but supported, with "WordPress Address" already using HTTPS though**
![Screen Shot 2021-01-28 at 12 17 55 PM](https://user-images.githubusercontent.com/3531426/106194416-a8685b00-6163-11eb-99eb-4356ad38b5a8.png)

**HTTPS set up, but not supported (i.e. there are problems)**
![Screen Shot 2021-01-28 at 12 15 33 PM](https://user-images.githubusercontent.com/3531426/106194455-b8803a80-6163-11eb-9c4e-9b6ddba3a497.png)

@TimothyBJacobs One question you might know the answer to: Now that I've made the test asynchronous and added a REST endpoint for it, do I need to update `wp-api-generated.js`? At what point is that generated and how?

felixarntz commented:
Committed via https://core.trac.wordpress.org/changeset/50072.
